Improve documentation (#64)
elastic/elastic-client-generator: README.md

# elastic client generator

This repository contains the Elasticsearch request/response definitions in TypeScript,
you can find them inside `/specification/specs`.
The `specification` folder contains a TypeScript program that compiles the entire definition
in a custom JSON representation that can be used for generating other language clients.

## How to generate the JSON representation

```
# clone the project
$ git clone https://github.com/elastic/elastic-client-generator.git

# install the dependencies
$ npm install --prefix specification

# generate the JSON representation
$ npm run generate-schema --prefix specification

# the generated output can be found in ./output/schema/schema.json
$ cat output/schema/schema.json
```

### Structure of the JSON representation

The representation is a JSON object with two top level keys:

> ```json

Author: Cool!

friendly specification, we do offer a set of aliases based on which string we do expect, for exmaple:

```ts
type ScrollId = string
type ScrollIds = string
type CategoryId = string
type ActionIds = string
type Field = string
type Id = string | number
type Ids = string | number | string[]
type IndexName = string
type Indices = string | string[]
...
```

You can find the full list [here](https://github.com/elastic/elastic-client-generator/blob/update-docs/specification/specs/common.ts),
feel free to add more if it feels appropriate!

### Dates

The `Date` type in TypeScript refers to the JavaScript `Date` object,
since Elasticsearch needs a string or a numeric value, there are aliases also for date types:

```ts
type Timestamp = string
type TimeSpan = string
interface Date {}
```

## FAQ

### A specific property is not always present, how do I define it?

When you define a property the syntax is `propertyName: propertyType`.
By default a property is required to exists. If you know that a property will not
always be there, you can add a question mark juist before the column:

```ts
propertyRequired: string
propertyOptional?: string
```

> ### A defintion is missing, how do I add it?

Reviewer: Typo

```suggestion
### A definition is missing, how do I add it?
```

> All the definitons are inside `specifications/specs` folder, you should explore its content and find the

Reviewer: Moar links (and  typos)

```suggestion
All the definitons are in the [`specifications/specs`](specifications/specs) folder, you should explore its content and find the
```

most approriate place where to add the new defintion. You can either create a new file or update an existing one.
If possible, try to reuse existing type definitions (eg `Indices` instead of `string`).

### A definition is not correct, how do I fix it?

All the definitons are inside `specifications/specs` folder, search the bad defintion and update it,
you can find above how to run the validation of the spec.

### An endpoint is missing, how do I add it?

All the endpoint definitons are inside `specifications/specs/json` folder, which contains a series of
JSON files taken directly from the Elasticsearch rest-api-spec.
You should copy from there the missing endpoint and add it here.

### An endpoint definition is not correct, how do I fix it?

All the endpoint definitons are inside `specifications/specs/json` folder, which contains a series of
JSON files taken directly from the Elasticsearch rest-api-spec.
You should copy from there the updated endpoint defintion and change it here.

### The validation in broken on GitHub but works on my machine!

> Very likely thew recordings in our machine are stale, you can regenerate them

Reviewer: Typos

```suggestion
Very likely the recordings on your machine are stale, you can regenerate them
```

by executing the following command (it will take a while).

```sh
PULL_LATEST=true ./run-validations
```

### Which editor should I use?

Any editor is fine, but to have a better development experience it should be configured

> to work with TypeScript. If you don't have it, [Visual Studio Code](https://code.visualstudio.com/)
> comes with TypeScript support out of the box.

Reviewer: Also mention IntelliJ that most Java developers (i.e. ES devs) use.

```suggestion
to work with TypeScript. [Visual Studio Code](https://code.visualstudio.com/) and
[IntelliJ IDEA](https://www.jetbrains.com/idea/) come with TypeScript support out of the box.
```
